ring-devel/src/ring/middleware/reload.clj

```diff
@@ -4,19 +4,35 @@
   This middleware should be limited to use in development environments."
   (:require [ns-tracker.core :refer [ns-tracker]]))
 
+(defn- reloader [dirs retry?]
+  (let [modified-namespaces (ns-tracker dirs)
+        load-queue (java.util.concurrent.LinkedBlockingQueue.)]
+    (fn []
+      (locking load-queue
+        (doseq [ns-sym (modified-namespaces)]
+          (.offer load-queue ns-sym))
+        (loop []
+          (when-let [ns-sym (.peek load-queue)]
+            (if retry?
+              (do (require ns-sym :reload) (.remove load-queue))
+              (do (.remove load-queue) (require ns-sym :reload)))
+            (recur)))))))
+
 (defn wrap-reload
   "Reload namespaces of modified files before the request is passed to the
   supplied handler.
 
   Accepts the following options:
 
-  :dirs - A list of directories that contain the source files.
-          Defaults to [\"src\"]."
+  :dirs                   - A list of directories that contain the source files.
+                            Defaults to [\"src\"].
+  :reload-compile-errors? - If true, keep attempting to reload namespaces
+                            that have compile errors.  Defaults to true."
   {:arglists '([handler] [handler options])}
-  [handler & [options]]
-  (let [source-dirs (:dirs options ["src"])
-        modified-namespaces (ns-tracker source-dirs)]
+  [handler & {:keys [dirs reload-compile-errors?]
```

Comment:
I think you've gotten the syntax wrong a little here. This will expand out the map, because you've lost the containing vector. Instead:

```clojure
[handler & [{:keys [dirs reload-compile-errors?]
             :or   {dirs ["src"], reload-compile-errors? true}}]]
```

Incidentally, in a later patch I'm going to get rid of the `& [options]` syntax because while it's a nice shortcut, it also messes up the function's arity.